src/monodroid/jni/monodroid-glue.c

```diff
@@ -704,55 +737,192 @@ typedef void* (*mono_mkbundle_init_ptr) (void (*)(const MonoBundledAssembly **),
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
+}
+
+static void*
+load_dso_from_override_dirs (const char *name, int dl_flags)
+{
+	if (name == NULL)
+		return NULL;
+
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
+		if (override_dirs [i] == NULL)
+			continue;
+
+		mono_bool needs_free = FALSE;
+		char *full_name = get_full_dso_path (override_dirs [i], name, &needs_free);
+		void *handle = load_dso (full_name, dl_flags, FALSE);
+		if (needs_free)
+			free (full_name);
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
 }
 
 static char*
-get_bundled_app (JNIEnv *env, jstring dir)
+get_existing_dso_path_on_disk (const char *base_dir, const char *dso_name, mono_bool *needs_free)
 {
-	const char *v;
-	char *libapp;
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
 
-#ifndef RELEASE
-	libapp = path_combine (override_dirs [0], "libmonodroid_bundle_app.so");
+	*needs_free = FALSE;
+	char *dso_path = get_full_dso_path (base_dir, dso_name, needs_free);
+	if (file_exists (dso_path))
+		return dso_path;
 
-	if (file_exists (libapp))
-		return libapp;
+	*needs_free = FALSE;
+	free (dso_path);
+	return NULL;
+}
 
-	free (libapp);
-#endif
+static void
+__dso_alloc_cleanup (char **dso_path, mono_bool *needs_free)
+{
+	if (dso_path != NULL) {
+		if (needs_free && *needs_free)
+			free (*dso_path);
+		*dso_path = NULL;
+	}
 
-	if (dir) {
-		v = (*env)->GetStringUTFChars (env, dir, NULL);
-		if (v) {
-			libapp = path_combine (v, "libmonodroid_bundle_app.so");
-			(*env)->ReleaseStringUTFChars (env, dir, v);
-			if (file_exists (libapp))
-				return libapp;
-		}
+	if (needs_free)
+		*needs_free = FALSE;
+}
+
+static char*
+get_full_dso_path_on_disk (const char *dso_name, mono_bool *needs_free)
+{
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
+
+	*needs_free = FALSE;
+	if (embedded_dso_mode)
+		return NULL;
+
+	char *dso_path = NULL;
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
+		if (override_dirs [i] == NULL)
+			continue;
+		dso_path = get_existing_dso_path_on_disk (override_dirs [i], dso_name, needs_free);
+		if (dso_path != NULL)
+			return dso_path;
+		__dso_alloc_cleanup (&dso_path, needs_free);
+	}
+
+	for (int i = 0; i < app_lib_directories_size; i++) {
+		dso_path = get_existing_dso_path_on_disk (app_lib_directories [i], dso_name, needs_free);
+		if (dso_path != NULL)
+			return dso_path;
+		__dso_alloc_cleanup (&dso_path, needs_free);
 	}
+
 	return NULL;
 }
 
+// This function could be improved if we somehow marked an apk containing just the bundled app as
+// such - perhaps another __XA* environment variable? Would certainly make code faster.
+static void
+setup_bundled_app (const char *dso_name)
+{
+	int dlopen_flags = RTLD_LAZY;
+	void *libapp = NULL;
+
+	if (embedded_dso_mode) {
+		log_info (LOG_DEFAULT, "bundle app: embedded DSO mode");
+		libapp = load_dso_from_override_dirs (dso_name, dlopen_flags);
```

Comment:
I don't understand what this is *for*.  As per comments elsewhere, `load_dso_from_override_dirs()` should always return NULL in `Release` configuration builds, and "mkbundled apps" is generally of questionable use in `Debug` configuration builds -- it "kills" fast deployment, requires that the `.apk` always be rebuilt & redeployed, etc. -- so why would we want to look for `libmonodroid_bundle_app.so` in the override dirs when `embedded_dso_mode==TRUE`?